eth/filters/filter.go

```diff
@@ -129,6 +129,32 @@ func (f *Filter) Logs(ctx context.Context) ([]*types.Log, error) {
 	if f.end == rpc.LatestBlockNumber.Int64() || f.end == rpc.PendingBlockNumber.Int64() {
 		end = head
 	}
+	if f.begin == rpc.FinalizedBlockNumber.Int64() || f.end == rpc.FinalizedBlockNumber.Int64() {
```

Comment:
Looks like a lot of repitition -- also, I imagine we need to do similar things in other places in the code. What do you think of doing something like the below? Then it might be more amenable to reuse somewhere else, and possibly also incorporate so that `pending` and `latest` are treated similarly. 

```golang
	resolveSpecial := func(number int64) (int64, error) {
		var hdr *types.Header
		switch number {
		case rpc.FinalizedBlockNumber.Int64():
			hdr, _ := f.sys.backend.HeaderByNumber(ctx, rpc.FinalizedBlockNumber)
			if hdr == nil {
				return 0, errors.New("finalized header not found")
			}
		case rpc.SafeBlockNumber.Int64():
			hdr, _ := f.sys.backend.HeaderByNumber(ctx, rpc.SafeBlockNumber)
			if hdr == nil {
				return 0, errors.New("safe header not found")
			}
		}
		return hdr.Number.Int64(), nil
	}
	if f.begin, err = resolveSpecial(f.begin); err != nil {
		return nil, err
	}
	if f.end, err = resolveSpecial(f.end); err != nil {
		return nil, err
	}
```

Reply:
Oh ya I like that approach, I was trying to come up w/ something similar that was still readable but your approach is much better than what I had. :). I'll make that change.

Reply:
Hmm, actually there's a couple problems w/ that approach:

- we need to set `end` not `f.end`, and more annoyingly `end` is uint64 while `f.end` is int64.
- we also need to make sure we don't overwrite `f.begin` or `end` values set earlier by the pending/latest if statements.

I'll see if I can come up w/ a clean solution that makes it all easier to read and reuses the resolution code.

Reply:
Ok @holiman I force pushed an new version that works with a slightly modified version of what you suggested, and added test coverage for the finalized case (I didn't add safe coverage since that would require changes to testBackend since safe status isn't stored in the DB).  I'm happy with how it came out, I think it's overall cleaner than what we started with, so PTAL thanks!